tszip/compression.py

```diff
@@ -85,25 +86,26 @@ def compress(ts, path, variants_only=False, stdout=False):
         the location of the compressed file.
     :param bool variants_only: If True, discard all information not necessary
         to represent the variants in the input file.
-    :param bool stdout: If True, write the tree sequence to stdout.
     """
-    destination = str(path)
+    destination = path
     # Write the file into a temporary directory on the same file system so that
     # we can write the output atomically.
-    destdir = os.path.dirname(os.path.abspath(destination))
-    with tempfile.TemporaryDirectory(dir=destdir, prefix=".tszip_work_") as tmpdir:
+    # destdir = os.path.dirname(os.path.abspath(destination))
+    with tempfile.TemporaryDirectory(dir=None, prefix=".tszip_work_") as tmpdir:
```

Comment:
This is almost right I think, but it does break the atomic write contract which ensures that we don't leave partially written files lying around. The idea of putting the temporary file in the same directory as the destination is that moving the temporary file to it's final destination is simply a rename, which is as close to atomic as you can get. However, if the temporary file is in the standard temp directory, then this becomes a real move because ``/tmp`` is usually a in a different file system.

So, how about this:

```python
try:
    destination = pathlib.Path(path).resolve()
    is_path = True
    tmpdir = destination.parent # I think this is right?
except TypeError:
    is_path = False 
    tmpdir = None
with tempfile.TemporaryDirectory(dir=tmpdir, prefix=".tszip_work_") as tmpdir:
      # same as before 
      filename = pathlib.Path(tmpdir) / "tmp.trees.tgz"
      with zarr.ZipStore(filename, mode="w") as store:
          # blah blah
      if is_path:
          os.replace(..) # atomic file move
      else:
          # write chunks to dest
```
```

Reply:
Seems like a good idea to update to pathlib here as well while we're at it? That's no necessary, though.

Reply:
Great suggestion! This solution works really great and actually passes all of the tests that we've put forth.

There is a somewhat insidious bug when actually running on the command line: 
```
pip install .
tszip -c tests/files/1.0.0.trees
```
though because `sys.stdout` expects strings (but in the tests we overload `sys.stdout` to be a buffer in `wb+` mode) for capturing output in our pytest suite...

Reply:
That's weird - what happens when we write to stdout? It seems to work OK when we write binary data to stdout in msprime (try ``msp ancestry``) and it doesn't look like there's any changes made there... Ah, maybe because we're writing directly to the file descriptor there and not doing it in Python.

OK, trying ``sys.stdout.buffer`` as the argument?